### DISCUSSION TITLE
Use only one form for different relations

### DISCUSSION THREAD

**Discussion by guidot** [Budibase/budibase#7547] [Help]
I have three primary tables: _Tracker_, _Accounts_ and _Clients_.
And then there is a secondary table: _Questions_.
Each row in the primary tables can have one or more _questions_ attached. (The _question_ table has 3 relationship columns, one for each primary table).

When I show the row of one of the primary tables I also display the attached _questions_.
Buttons allow me to navigate to a screen for adding a new _question_ (or editing one of the already attached ones).

Independently from which of the primary tables I come I want to use only one screen for the creation of a _question_ (and one other for editing of an existing one). 

My question is: How can I save the correct relation? 

It was simple to do when I first built the question form inside the screen that displayed Tracker data (and changed it's visibility using conditions and states).
![grafik](https://user-images.githubusercontent.com/6181359/187738812-5b31fd24-7b5b-473f-8a33-a8b4d9e78d71.png)

This is the save action: 
![grafik](https://user-images.githubusercontent.com/6181359/187731904-d4043cb1-3719-4b7d-ae65-b6eaed57114d.png)
(`R_Tracker` is the relationship column)

But now I need to 
1. Pass on the information from which primary table the dialog is being opened. That should be no problem setting a variable with `Update State`.
2. Include a conditional that updates the correct relationship column.
    The save row action doesn't allow to set the column dynamically. So I need to save all three of them, `R_Tracker`, `R_Accounts` and `R_Clients`. Each time checking for the state variable from 1. How can I achieve that? Would something like 
    ```
    {{#if (eq State.relation "Tracker")}}
        {{ arrayify Tracker Repeater.Tracker._id }}
    {{/if}} 
    ``` 
    work?
3. How do I access the related tables? Do I need to add all three data providers or is there a better way?

**Comment by andz-bb** [comment]
Hey @guidot,

I've played around with this a bit and think I have a solution for you. 

![Screen Recording 2022-09-07 at 19 21 20](https://user-images.githubusercontent.com/110921612/188952411-5dc5b53c-4d7b-4b97-bcb2-a7715ee3bae6.gif)


Instead of using app state I found it best to use URL params to include the relation and ID in the route for the new question screen.

<img width="517" alt="Screenshot 2022-09-07 at 19 00 02" src="https://user-images.githubusercontent.com/110921612/188947117-2eb08e58-acec-41f5-a6ce-1f7dad408fe5.png">

For updating columns I used some custom JS for the Save Row action on the save button in the generic new question screen.

<img width="1586" alt="Screenshot 2022-09-07 at 19 01 08" src="https://user-images.githubusercontent.com/110921612/188947280-98b99d81-bb7a-4f0f-9cc0-af1c52b42dd8.png">

The JS checks the relation type from the param on the URL and then returns either the relevant ID or null. Here's the script I used for the value of the Tracker column:

`return $("URL.relation") === 'tracker' ? $("URL.relationid") : null;`

Hopefully this is what you were looking for! If you have any questions about what I am describing here let me know and I'll be happy to go into more detail. 😊

**Comment by guidot** [comment]
Hi @andz-bb,
Thanks for your help! Cool, I didn't know about the URL params, way more elegant than using state.

I have ~~some~~ one more unresolved issues with this:

EDIT: 1. and 2. were overcomplicated thinking. Please ignore.

~~1. The relations are one (primary tables) to many (question table).
To correctly store the relations I assume I need to have a second `Save Row` action to save the question ID in the related primary table. I tried `arrayify` as suggested in another discussion:
(See below for the structure I used here.) But I think I didn't find the correct variable that holds the new question's ID. Using the above doesn't work.~~

~~2. How do I structure my data provider? My first attempt looked like this:
  But that forces me to repeat several of the elements. Not very elegant.
  So how about nesting them?~~

~~Now I could use the above JS snippet. But how do I amend it to save my one-to-many relations?~~

In the _New question for tracker/client/account_ form I also need to display the content of another column (the name) in that related tracker/client/account row. How would I access it using the URL param approach?

**Comment by elysium27** [comment]
Hi @andz-bb ,

Could you export the app? 

I am trying to do the same thing with the OKR template but unable to do so. 

![image](https://user-images.githubusercontent.com/83823695/189314986-f7aa3cde-2118-4a95-8531-674bc6daa41c.png)

**Reply by andz-bb** [reply]
Hi @elysium27, here's that app export. Hope it helps!
[7547-export-1662969964268.txt](https://github.com/Budibase/budibase/files/9546057/7547-export-1662969964268.txt)

If there's anything you're not sure about with it let me know and I can have a look 😊

**Reply by guidot** [reply]
Hi @andz-bb,

Many thanks for sharing! :)

Can you explain, please: What's the purpose of the TRACKERID, ACCOUNTID and CLIENTID columns in the questions table? I didn't find them referenced anywhere.
However, looking at the tracker:Id screen I saw the table inside the Question Data Provider didn't show the related questions. The right hand side of the Question Data Provider _Filter_ was set to  `Null`. After I set this to `TrackerId` the table worked as expected. Was that the idea of those columns?
I'm asking because they don't seem to be necessary for that. When you place the Question Data Provider inside the Tracker Repeater you can set _Data_ to the relation `Repeater.Tracker.Questions` and don't need a filter at all.
